haskell-tests/Generators.hs

```diff
@@ -0,0 +1,977 @@
+    return RawUpdateInstruction{..}
+
+genLevel2RawUpdateInstruction :: SChainParametersVersion cpv -> Gen RawUpdateInstruction
+genLevel2RawUpdateInstruction scpv = do
+    ruiSeqNumber <- Nonce <$> arbitrary
+    ruiEffectiveTime <- oneof [return 0, TransactionTime <$> arbitrary]
+    ruiTimeout <- TransactionTime <$> arbitrary
+    ruiPayload <- genLevel2UpdatePayload scpv
+    return RawUpdateInstruction{..}
+
+-- |Generate an 'Authorizations' structure and the list of key pairs.
+-- The threshold for each access structure is specified.
+genAuthorizationsAndKeys ::
+    forall cpv.
+    IsChainParametersVersion cpv =>
+    -- |Threshold for each access structure
+    UpdateKeysThreshold ->
+    Gen (Authorizations cpv, [KeyPair])
+genAuthorizationsAndKeys thr = do
+    let nKeys = case chainParametersVersion @cpv of
+            SCPV0 -> fromIntegral thr * 12
+            SCPV1 -> fromIntegral thr * 14
+    kps <- vectorOf nKeys genSigSchemeKeyPair
+    let asKeys = Vec.fromList $ correspondingVerifyKey <$> kps
+    let genAccessStructure = do
+            asnKeys <- choose (fromIntegral thr, nKeys)
+            accessPublicKeys <- Set.fromList . take asnKeys <$> shuffle [0 .. fromIntegral nKeys - 1]
+            return AccessStructure{accessThreshold = thr, ..}
+    asEmergency <- genAccessStructure
+    asProtocol <- genAccessStructure
+    asParamElectionDifficulty <- genAccessStructure
+    asParamEuroPerEnergy <- genAccessStructure
+    asParamMicroGTUPerEuro <- genAccessStructure
+    asParamFoundationAccount <- genAccessStructure
+    asParamMintDistribution <- genAccessStructure
+    asParamTransactionFeeDistribution <- genAccessStructure
+    asParamGASRewards <- genAccessStructure
+    asBakerStakeThreshold <- genAccessStructure
+    asAddAnonymityRevoker <- genAccessStructure
+    asAddIdentityProvider <- genAccessStructure
+    asCooldownParameters <- justForCPV1A genAccessStructure
+    asTimeParameters <- justForCPV1A genAccessStructure
+    return (Authorizations{..}, kps)
+
+genLevel1Keys ::
+    UpdateKeysThreshold ->
+    Gen (HigherLevelKeys Level1KeysKind, [KeyPair])
+genLevel1Keys thr = do
+    kps <- vectorOf (fromIntegral thr * 2) genSigSchemeKeyPair
+    let hlkKeys = Vec.fromList $ correspondingVerifyKey <$> kps
+    return (HigherLevelKeys{hlkThreshold = thr, ..}, kps)
+
+genRootKeys ::
+    UpdateKeysThreshold ->
+    Gen (HigherLevelKeys RootKeysKind, [KeyPair])
+genRootKeys thr = do
+    kps <- vectorOf (fromIntegral thr * 2) genSigSchemeKeyPair
+    let hlkKeys = Vec.fromList $ correspondingVerifyKey <$> kps
+    return (HigherLevelKeys{hlkThreshold = thr, ..}, kps)
+
+genKeyCollection :: IsChainParametersVersion cpv => UpdateKeysThreshold -> Gen (UpdateKeysCollection cpv, [KeyPair], [KeyPair], [KeyPair])
+genKeyCollection thr = do
+    (rootKeys, a) <- genRootKeys thr
+    (level1Keys, b) <- genLevel1Keys thr
+    (level2Keys, c) <- genAuthorizationsAndKeys thr
+    return (UpdateKeysCollection{..}, a, b, c)
```

Comment:
This is a poor quality module. If refactoring was desired, then documentation should accompany it. As it is, it is a massive module with some generators, no clarity on the quality of these generators, or their coverage. 

While that was also true before this move, for refactoring to have value it should improve maintainability of code. This does not achieve that.